### DISCUSSION TITLE
Retrieving the Navigator Language

### DISCUSSION THREAD

**Discussion by PoshoDev** [zauberzeug/nicegui#2771] [Q&A] [answered]
### Question

Hello there, :smile: 

I need some guidance on how to retrieve the `navigator.language` or `navigator.userLanguage` so I can display my niceGUI site in different ways depending on the user's language. Something like:

```python
@ui.page("/"):
def mypage():
    if get_language() == "en-US":
        # blah blah blah...
    else:
        #blah blah blah...
```

The closest I've gotten is this approach:
```python
from nicegui import ui

@ui.page("/1")
def this_works():
    async def get_lang():
        js_code = 'navigator.language || navigator.userLanguage'
        language = await ui.run_javascript(js_code)
        ui.notify(f"Your language is {language}")

    ui.button("Get language", on_click=get_lang)
```
![image](https://github.com/zauberzeug/nicegui/assets/32964490/168d0bb4-8d74-4843-962d-6ae978f99b35)

The problem with this is that an user has to interact with the button (or a NiceGUI element) in order to retrieve the language, so that's not the best for my use case.

I have attempted other approaches, like just calling `get_lang()` like a normal function, but then the page function would need to be `async` and apparently you can't do that.

Another approach I tried was using a similar implementation as [another issue I had](https://github.com/zauberzeug/nicegui/discussions/2405#discussioncomment-8190904) (#2405) solved by @falkoschindler:

```python
@ui.page("/3")
def this_attempt_3():
    ui.add_head_html('''
        <script>
            function get_lang() {
                if (document.readyState === 'complete') {
                    emitEvent('lang_event', {
                        language: navigator.language || navigator.userLanguage
                    });
                } else {
                    setTimeout(get_lang, 100);
                }
            }
            get_lang();
        </script>
    ''')
    ui.on('lang_event', lambda e: print(e.args))
```

This works, but only after the page has been fully-loaded, so the original desired implementation can't be used. You might say _"just remove the `if (document.readyState === 'complete')` condition"_, but then the event is not called at all.

<details>
<summary>Here are other attempts I tried.</summary>

```python
ui.add_head_html('''
    <script>
    document.addEventListener('DOMContentLoaded', function() {
        emitEvent('language', {
            language: navigator.language || navigator.userLanguage
        });
    });
    </script>
''')
ui.on('language', lambda e: print(e.args))
```

```python
ui.add_head_html('''
    <script>
        emitEvent('lang3', {
            lang: navigator.language || navigator.userLanguage
        });
        lang3f();
    </script>
''')
ui.on('lang3', lambda e: print("lol", e.args))
```

```python
ui.add_head_html('''
    <script>
        function lang4f() {
            if (typeof emitEvent === 'function') {
                emitEvent('lang4', {
                    lang4sss: navigator.language || navigator.userLanguage
                });
            } else {
                setTimeout(lang4f, 100);
            }
        }
        lang4f();
    </script>
''')
ui.on('lang4', lambda e: print(e.args))
```

</details>

As if it wasn't obvious, my JavaScript is not the best. So I thank anyone who'd be willing to give me some guidance and patience. :pray:

**Comment by rodja** [comment, marked as answer]
Running JavaScript is only possible after the client has connected (eg. page is loaded). See https://nicegui.io/documentation/page#wait_for_client_connection on how you could do this.

**Reply by PoshoDev** [reply]
I see. Using `async` functions I can get this done. Thank you so much! :smile:
